src/backend/renderer/element/mod.rs

```diff
@@ -0,0 +1,1355 @@
+    };
+    ($(#[$attr:meta])* $vis:vis $name:ident<$renderer:ident, $($custom:ident),+> $(where $($target:ty: $bound:tt $(+ $additional_bound:tt)*),+)?; $($tail:tt)*) => {
+        $crate::render_elements_internal!(@enum $(#[$attr])* $vis $name<$renderer, $($custom),+>; $($tail)*);
+        $crate::render_elements_internal!(@impl $name<$renderer, $($custom),+> $(where $($target: $bound $(+ $additional_bound)*),+)?; $($tail)*);
+    };
+    ($(#[$attr:meta])* $vis:vis $name:ident<$renderer:ident> $(where $($target:ty: $bound:tt $(+ $additional_bound:tt)*),+)?; $($tail:tt)*) => {
+        $crate::render_elements_internal!(@enum $(#[$attr])* $vis $name<$renderer>; $($tail)*);
+        $crate::render_elements_internal!(@impl $name<$renderer> $(where $($target: $bound $(+ $additional_bound)*),+)?; $($tail)*);
+        $crate::render_elements_internal!(@from $name<$renderer>; $($tail)*);
+    };
+    ($(#[$attr:meta])* $vis:vis $name:ident; $($tail:tt)*) => {
+        $crate::render_elements_internal!(@enum $(#[$attr])* $vis $name; $($tail)*);
+        $crate::render_elements_internal!(@impl $name; R; $($tail)*);
+        $crate::render_elements_internal!(@from $name; $($tail)*);
+    };
+}
+
+pub use render_elements;
+
+/// New-type wrapper for wrapping owned elements
+/// in render_elements!
+#[derive(Debug)]
+pub struct Wrap<C>(C);
+
+impl<C> From<C> for Wrap<C> {
+    fn from(from: C) -> Self {
+        Self(from)
+    }
+}
+
+impl<R, C> RenderElement<R> for Wrap<C>
+where
+    R: Renderer,
+    C: RenderElement<R>,
+{
+    fn id(&self) -> &Id {
+        self.0.id()
+    }
+
+    fn current_commit(&self) -> CommitCounter {
+        self.0.current_commit()
+    }
+
+    fn location(&self, scale: Scale<f64>) -> Point<i32, Physical> {
+        self.0.location(scale)
+    }
+
+    fn src(&self) -> Rectangle<f64, BufferCoords> {
+        self.0.src()
+    }
+
+    fn transform(&self) -> Transform {
+        self.0.transform()
+    }
+
+    fn geometry(&self, scale: Scale<f64>) -> Rectangle<i32, Physical> {
+        self.0.geometry(scale)
+    }
+
+    fn draw(
+        &self,
+        renderer: &mut R,
+        frame: &mut <R as Renderer>::Frame,
+        location: Point<i32, Physical>,
+        scale: Scale<f64>,
+        damage: &[Rectangle<i32, Physical>],
+        log: &slog::Logger,
+    ) -> Result<(), <R as Renderer>::Error> {
+        self.0.draw(renderer, frame, location, scale, damage, log)
+    }
+
+    fn damage_since(
+        &self,
+        scale: Scale<f64>,
+        commit: Option<CommitCounter>,
+    ) -> Vec<Rectangle<i32, Physical>> {
+        self.0.damage_since(scale, commit)
+    }
+
+    fn opaque_regions(&self, scale: Scale<f64>) -> Vec<Rectangle<i32, Physical>> {
+        self.0.opaque_regions(scale)
+    }
+
+    fn underlying_storage(&self, renderer: &R) -> Option<UnderlyingStorage<'_, R>> {
+        self.0.underlying_storage(renderer)
+    }
+}
+
+#[cfg(all(test, feature = "renderer_gl"))]
+#[allow(dead_code)]
+mod tests {
```

Comment:
This file is getting kind of long. We should probably move these tests to their own file.